mlflow/data/numpy_dataset.py

```diff
@@ -0,0 +1,140 @@
+
+        md5 = hashlib.md5()
+
+        # hash trimmed feature contents
+        try:
+            md5.update(pd.util.hash_array(trimmed_features))
+        except TypeError:
+            md5.update(np.int64(trimmed_features.size))
+        # hash full feature dimensions
+        for x in self._features.shape:
+            md5.update(np.int64(x))
+
+        # hash trimmed targets contents
+        if self._targets:
+            flattened_targets = self._targets.flatten()
+            trimmed_targets = flattened_targets[0:MAX_ROWS]
+            try:
+                md5.update(pd.util.hash_array(trimmed_targets))
+            except TypeError:
+                md5.update(np.int64(trimmed_targets.size))
+            # hash full feature dimensions
+            for x in self._targets.shape:
+                md5.update(np.int64(x))
+
+        return md5.hexdigest()[:8]
+
+    def _to_dict(self, base_dict: Dict[str, str]) -> Dict[str, str]:
+        """
+        :param base_dict: A string dictionary of base information about the
+                          dataset, including: name, digest, source, and source
+                          type.
+        :return: A string dictionary containing the following fields: name,
+                 digest, source, source type, schema (optional), profile
+                 (optional).
+        """
+        base_dict.update(
+            {
+                "schema": json.dumps({"mlflow_tensorspec": self.schema.to_dict()}),
+                "profile": json.dumps(self.profile),
+            }
+        )
+        return base_dict
+
+    @property
+    def source(self) -> FileSystemDatasetSource:
+        """
+        TODO: Numpy docs
+        """
+        return self._source
+
+    @property
+    def features(self) -> Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]]:
+        return self._features
+
+    @property
+    def targets(self) -> Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]]:
+        return self._targets
+
+    @property
+    def profile(self) -> Optional[Any]:
+        """
+        TODO: Numpy docs
+        """
+        return {
+            "shape": self._features.shape,
+        }
+
+    @property
+    def schema(self) -> Schema:
+        """
+        An MLflow TensorSpec schema representing the tensor dataset
+        """
+        # TODO: Error handling
+        return _infer_schema(self._features)
+
+
+def from_numpy(
+    features: Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]],
+    source: str,
+    targets: Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]] = None,
+    name: Optional[str] = None,
+    digest: Optional[str] = None,
+) -> NumpyDataset:
+    """
+    :param features: NumPy features, represented as an np.ndarray, list of np.ndarrays
+                    or dictionary of named np.ndarrays.
+    :param source: The source from which the NumPy data was derived, e.g. a filesystem
+                    path, an S3 URI, an HTTPS URL etc.
```

Comment:
Can we more specifically say that the source needs to refer to a local filesystem or cloud blob storage location?